numpy/core/src/multiarray/dragon4.c

```diff
@@ -42,6 +42,18 @@
 #define DEBUG_ASSERT(stmnt) do {} while(0)
 #endif
 
+static inline npy_uint64
+bitmask_u64(npy_uint32 n)
+{
+    return ~(~((npy_uint64)0) << n);
+}
+
+static inline npy_uint32
+bitmask_u32(npy_uint32 n)
+{
+    return ~(~((npy_uint32)0) << n);
+}
```

Comment:
ok, but still it seems to violate KISS to convert 0x7 into a function call, even if the compiler is smart enough to make it a constant.

Reply:
`0x7FFFFF` is more clearly 23 bits as `bitmask_u32(23)` - where do you draw the line? Isn't it better to just be consistent? Perhaps the function just needs a better name? `lower_n_bits`?

Reply:
The point is you're not supposed to be translating to `0x7F`, because `0x7F` isn't a useful way to think about it in the first place. `x & bitmask(n)` reads "take the first N bits of x`. Ideally we'd just use bitfields to encode this

```
struct ieee_float {
    uint64_t sign: 1;
    uint64_t mantissa: 23;
    uint64_t exponent: 8;
}
```

That's not cross-platform though. Using `bitmask` at least allows the numbers 1, 23, and 8 to appear in the source code.

Reply:
Just to jump in, I think `bitmask_u32(3)` is more accessible than `0x7` - the latter doesn't make clear we are trying to select the last 3 bits, and would be inconsistent with the rest of the file.